### DISCUSSION TITLE
Can't get Modbus RTU data from an Arduino Server via Modbus serial

### DISCUSSION THREAD

**Discussion by nikstp** [pymodbus-dev/pymodbus#1099] [Help] [answered]
Hi, 

PyModbus version: 3.0.0.5dev

I've been trying to verify that I can at least connect to an Arduino Uno acting as a Modbus RTU server via a USB-serial cable with an async client, but I had issues with opening the COM port, so I've switched to the sync version. I can open the serial port now, but I don't see any data coming in and I get the following error:

_Modbus Error: [Input/Output] No Response received from the remote unit/Unable to decode response_

I've verified with Modbus Poll that the Arduino Uno does act as a server - I can read from and write to it with MP. I'd like to be able to do the same with PyModbus.

Any ideas on what I can check or change are welcome.  I've checked the baudrate, bytesize, parity, stopbits, and port multiple times.  

My code is as follows:

```
def handleHoldingRegisters(client):
    """Read/write holding registers."""
    print("In holding register function!")
    #read 8 holding registers at 0x42
    rr = client.read_holding_registers(1, 8, slave=0x42)
    #assert not rr.isError()
    print(str(rr))
    return rr


def setupSyncClient(clientPort):
    """Run client setup."""
    
    client = ModbusSerialClient(
            port = clientPort,
            # Common optional paramers:
            framer=ModbusRtuFramer,
            #timeout=3,
            retries=3,
            #retry_on_empty=False,
            #close_comm_on_error=True,
            #strict=False,
            # Serial setup parameters
            baudrate=9600,
            bytesize=8,
            parity="N",
            stopbits=1,
            #handle_local_echo=False,
        )
    _logger.info("### Create client object")

    return client


def runSyncModbusClient(client, modbus_calls):
    """Run sync client."""
    print("Client starting")
    client.connect()
    if modbus_calls:
        modbus_calls(client)
    client.close()

def main():
    print(f"started main at {time.strftime('%X')}")
    
    modbusTestClient = setupSyncClient("COM51")

    while True:
        runSyncModbusClient(modbusTestClient, modbus_calls=handleHoldingRegisters)  
    
    
```

**Comment by janiversen** [comment]
Your com port looks wrong, it is normally something like /dev/com51. Apart from that you can activate debug and get more information.

**Reply by nikstp** [reply]
I'm on Windows, so it has to be "COM51" ior "com51". When I try "/dev/com51" or "/dev/COM51" I get the following error: 

_ERROR    serial         :221      could not open port '/dev/com51': FileNotFoundError(2, 'The system cannot find the path specified.', None, 3)_

How do you properly enable debug logging?

**Reply by janiversen** [reply]
Well if async cannot open the port then sync should have the same problem but might report it later and differently.

logging.setlevel()

**Reply by nikstp** [reply]
Okay, thank you, letsee if I can get anything useful out of the debug info.

**Reply by nikstp** [reply]
Now, I get the following:

It seems as if the port is open and able to send a command, but unless I check it with a logic analyzer, then I'm not 100% sure. What do you think?

2022-09-28 10:14:01,305 INFO     modbusClient   :110      ### Client starting
Client starting
In holding register function!
2022-09-28 10:14:01,312 INFO     modbusClient   :71       ### write holding register and read holding registers
2022-09-28 10:14:01,312 DEBUG    transaction    :130      Current transaction state - TRANSACTION_COMPLETE
2022-09-28 10:14:01,313 DEBUG    transaction    :134      Running transaction 18
2022-09-28 10:14:01,313 DEBUG    transaction    :291      SEND: 0x0 0x3 0x0 0x0 0x0 0x8 0x45 0xdd
2022-09-28 10:14:01,313 DEBUG    rtu_framer     :280      Changing state to IDLE - Last Frame End - None, Current Time stamp - 1664378041.31336
2022-09-28 10:14:01,319 DEBUG    base           :194      New Transaction state "SENDING"
2022-09-28 10:14:01,341 DEBUG    transaction    :312      Changing transaction state from "SENDING" to "WAITING FOR REPLY"
2022-09-28 10:14:04,342 DEBUG    transaction    :422      {msg_start} received, Expected 21 bytes Received 0 bytes !!!!
2022-09-28 10:14:04,343 DEBUG    transaction    :428      Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
2022-09-28 10:14:04,343 DEBUG    transaction    :327      RECV:
2022-09-28 10:14:04,344 DEBUG    rtu_framer     :249      Frame - [b''] not ready
2022-09-28 10:14:04,344 DEBUG    transaction    :528      Getting transaction 0
2022-09-28 10:14:04,345 DEBUG    transaction    :234      Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
Modbus Error: [Input/Output] No Response received from the remote unit/Unable to decode response

**Reply by janiversen** [reply]
yes it seems to be sending but your device does not respond. why not debug on the arduino…it might be that it 

a) does receive the message but do not respond
b) expect a different framer like e.g. ascii

**Reply by nikstp** [reply]
@janiversen okay, I'll do that as well. I'm relatively certain it expects an RTU framer, as within MP you can choose to send your frames as either RTU or ASCII and ASCII does not work, while RTU does as expected (the Arduino server is configured as RTU). It might be the case that it's not replying though, for some reason.

**Comment by janiversen** [comment]
are you sure the slave address is correct ? you can try with slave=0 which is broadcast.

**Reply by nikstp** [reply]
Yes, I've tried with 0x0, 0x41, 0x42, and 0x43 just because I'm not sure whether in PyModubs the address of the slave starts with 0 or 1. None work, I get the error as described in my first post. With ModbusPoll I can for sure access the slave over 0x42.

**Reply by janiversen** [reply]
it is not a matter of how pymodbus interpret the slave address, it simply passes it on unchanged, so if 0x42 works with other software then it will also work with pymodbus.

To me, it still sounds as if your problem is the port, you can also try with an older pymodbus version like 2.5.3 if you suspect the library have a problem.

**Reply by nikstp** [reply]
I'll have to create a separate venv from my curent project and try the async/sync client examples as is with the 2.5.3. just so we can either rule it out or if it works, focus on the current version.

**Comment by nikstp** [comment, marked as answer]
Okay, the issue was actually very simple (and silly) - 

The slave address is indeed 42, but putting 0x42 -> 66 in base 10, so I was trying to reach the wrong address. It should be just an int - number representing the slave ID.
